network/p2p/dns/cache.go

```diff
@@ -0,0 +1,88 @@
+package dns
+
+import (
+	"net"
+	"sync"
+	"time"
+)
+
+// defaultTimeToLive is the default duration a dns result is cached.
+const defaultTimeToLive = 5 * time.Minute
+
+// cache is a ttl-based cache for dns entries
+type cache struct {
+	sync.Mutex
```

Comment:
Agree and I initially did that. However, we are currently following a passive expiry of the cache entries. So, when we read a cache entry there is a chance that we also _invalidate_ it, which is a write operation. That is why a read lock does not fit the scenario. Of course, if we choose to have active expiry, we can segregate the read/write locks. However, as I mentioned in a [later comment](https://github.com/onflow/flow-go/pull/1086#discussion_r695054525) there is a tradeoff between going active on invalidating the cache entries. Looking forward to your inputs to start the refactoring.

Reply:
@yhassanzadeh13 Thanks for replying to the comment! :-)

> So, when we read a cache entry there is a chance that we also invalidate it, which is a write operation.

Would it not work to split that into two operations? E.g. first read the cache entry with the concurrent 'read' lock, and only if the entry is expired, then go ahead and lookup the domain outside of the lock, and finally use the blocking 'write' lock to update the entry?

Another idea to consider -- if we all agree not to respect the idea of per domain TTLs -- is to lookup expired domains after the fact. How would this work? We lookup the domain which is expired, but because there is an old cached record, we simply return that, and it is used for the connection. However, the DNS cache triggers its own async lookup which happens in parallel and/or later, and the DNS cache entry eventually gets updated, but not for that particular connection. This quick and dirty algorithm has the advantage of never letting the domain lookup add to the connection time (unless it's the very first connection).

Reply:
Per our conversation it needs async update of expired cache entries and I issued it up here to address on a separate PR: https://github.com/dapperlabs/flow-go/issues/5799